internal/data/message_writer.go

```diff
@@ -154,29 +165,104 @@ func (mw *messageWriter) writerLoop(index int) {
 				batch.messages = append(batch.messages, work.newMessage)
 			}
 			batch.data = append(batch.data, work.newData...)
-			batch.listeners = append(batch.listeners, work.result)
+			batch.listeners[*work.id] = work.result
 		case <-timeoutContext.Done():
 			timedOut = true
 		}
 
 		if batch != nil && (timedOut || (len(batch.messages)+len(batch.data) >= mw.conf.maxInserts)) {
 			batch.timeoutCancel()
-			err := mw.database.RunAsGroup(mw.ctx, func(ctx context.Context) error {
+			mw.persistMWBatch(batch)
+			batch = nil
+		}
+	}
+}
+
+func (mw *messageWriter) persistMWBatch(batch *messageWriterBatch) {
+	err := mw.database.RunAsGroup(mw.ctx, func(ctx context.Context) error {
+		return mw.writeMessages(ctx, batch.messages, batch.data)
+	})
+	if err != nil {
+		log.L(mw.ctx).Errorf("Failed batch message insert (pre-idempotency check): %s", err)
+		duplicatesRemoved := mw.removeIdempotencyDuplicates(mw.ctx, batch)
+		if duplicatesRemoved > 0 && (len(batch.messages) > 0 || len(batch.data) > 0) {
+			// We have a reduced scope batch to retry, with some duplicates removed
+			log.L(mw.ctx).Infof("Retrying batch insert after removing %d idempotency duplicates", duplicatesRemoved)
+			err = mw.database.RunAsGroup(mw.ctx, func(ctx context.Context) error {
 				return mw.writeMessages(ctx, batch.messages, batch.data)
 			})
-			for _, l := range batch.listeners {
-				l <- err
+		}
+	}
+	for _, l := range batch.listeners {
+		l <- err
+	}
+}
+
+func (mw *messageWriter) checkIdempotencyDuplicate(ctx context.Context, m *core.Message) error {
+	if m.IdempotencyKey != "" {
+		fb := database.MessageQueryFactory.NewFilter(ctx)
+		existing, _, err := mw.database.GetMessages(ctx, m.Header.Namespace, fb.Eq("idempotencykey", (string)(m.IdempotencyKey)))
+		if err != nil {
+			// Don't overwrite the original error for this - return -1 to the caller, who will return the previous error
+			log.L(mw.ctx).Errorf("Failed checking for idempotency errors: %s", err)
+			return nil
+		}
+		if len(existing) > 0 {
+			return i18n.NewError(ctx, coremsgs.MsgIdempotencyKeyDuplicateMessage, m.IdempotencyKey, existing[0].Header.ID)
+		}
+	}
+	return nil
+}
+
+func (mw *messageWriter) removeIdempotencyDuplicates(ctx context.Context, batch *messageWriterBatch) int {
+	duplicatesRemoved := 0
+	newMessageList := make([]*core.Message, 0, len(batch.messages))
+
+	// Spin through the messages in the batch, looking for existing messages that
+	for _, m := range batch.messages {
+		dupRemoved := false
+		idempotencyErr := mw.checkIdempotencyDuplicate(ctx, m)
+		if idempotencyErr != nil {
+			// We have an idempotency duplicate - we need to remove it from the batch so we can retry the rest
+			dupRemoved = true
+			if listener := batch.listeners[*m.Header.ID]; listener != nil {
+				// Notify the listener for this message,
+				listener <- idempotencyErr
 			}
-			batch = nil
+			delete(batch.listeners, *m.Header.ID)
+			// Remove all the data associated with this message from the batch
```

Comment:
Is there an edge case where two messages in the same batch could reference the same piece of data, and only one gets removed?